spec/message_spec.rb

```diff
@@ -0,0 +1,67 @@
+RSpec.describe "message encoding" do
+  # message Sally {
+  #   required string mood = 1;
+  #   optional uint32 age  = 2;
+  # }
+  class Sally
+    def initialize(attributes)
+      @attributes = {}
+      attributes.each do |attribute, value|
+        public_send("#{attribute}=", value)
+      end
+    end
+
+    ## Attributes
+
+    def age
+      @attributes[:age]
+    end
+
+    def age=(age)
+      raise ArgumentError, "age is not a number" unless age.respond_to?(:to_i)
+      raise ArgumentError, "age is too large" unless age.to_i <= 4_294_967_295
+      raise ArgumentError, "age cannot be negative" unless age.to_i >= 0
+      @attributes[:age] = age
+    end
+
+    def mood
+      @attributes[:mood]
+    end
+
+    def mood=(mood)
+      raise ArgumentError, "mood is not stringy" unless mood.respond_to?(:to_str)
+      @attributes[:mood] = mood
+    end
+
+    ## Encoding
+
+    def encode
+      stream = StringIO.new
+      stream.set_encoding(Encoding::BINARY)
+      encode_to_stream(stream)
+      stream.string
+    end
+
+    def encode_to_stream(stream)
```

Comment:
I like this idea for both encoding and decoding. There's a lot less digging around to figure out exactly how the encoding/decoding happens.

I have to keep reminding myself as I work on this code that the heuristics I normally use for designing objects are totally wrong when I am generating objects. This code doesn't have `maintenance` costs since it will always be re-generated. So I can sacrifice a lot more 'single responsibility' stuff in favor of 'readability stuff'.